sharktank/sharktank/layers/rotary_embedding_hf.py

```diff
@@ -0,0 +1,236 @@
+            indexing_maps = [
+                affine_map<(bs, sl, heads, two, halfdim) -> (bs, sl, heads, halfdim)>,
+                affine_map<(bs, sl, heads, two, halfdim) -> (bs, sl, heads, halfdim)>,
+                affine_map<(bs, sl, heads, two, halfdim) -> (bs, sl, heads, two, halfdim)>
+            ],
+            iterator_types = ["parallel", "parallel", "parallel", "parallel", "parallel"]
+        }
+        module {
+        util.func private @{{kernel_name}}(%x1: !x1,
+                                           %x2: !x2) -> !out {
+            %c0 = arith.constant 0 : index
+            %c1 = arith.constant 1 : index
+            %c2 = arith.constant 2 : index
+            %c3 = arith.constant 3 : index
+
+            %bs = tensor.dim %x1, %c0 : !x1
+            %sl = tensor.dim %x1, %c1 : !x1
+            %heads = tensor.dim %x1, %c2 : !x1
+            %halfdim = tensor.dim %x1, %c3 : !x1
+            %empty = tensor.empty(%bs, %sl, %heads, %halfdim) : !out
+
+            %out = linalg.generic #trait
+                   ins(%x1, %x2 : !x1, !x2)
+                   outs(%empty: !out) {
+                ^bb0(%xs1 : !dtype, %xs2 : !dtype, %o : !dtype):
+                %two_dim = linalg.index 3 : index
+                // Ideally, when the two dim is unrolled, this condition
+                // would become a no-op and we will not do any redundant
+                // computation.
+                %is_x1 = arith.cmpi eq, %two_dim, %c0 : index
+                %val = arith.select %is_x1, %xs1, %xs2 : !dtype
+                linalg.yield %val : !dtype
+            } -> !out
+            util.return %out : !out
+        }
+        }
+        """
+        return MLIRSpec(mlir)
+
+    return rope_select_concat
+
+
+select_concat = RoPEKernels()
+
+
+class RotaryEmbeddingHfLayer(BaseLayer):
+    """Computes a rotary embedding (RoPE)"""
+
+    def __init__(
+        self, *, head_dim: int, rope_theta: float = 10000.0, interleaved: bool = True
+    ):
+        super().__init__()
+        self.head_dim = head_dim
+        self.rope_theta = rope_theta
+        self.interleaved = interleaved
+
+    def _compute_theta(self):
+        # TODO: Add rope scaling.
+        dim = self.head_dim
+        # The original paper creates a d/2 dimensional space to represent
+        # the polar coordinates.
+        #
+        # From the original paper:
+        #   theta = 10000^{-2 (i - 1) / d}, i \in [1, 2, ..., d/2]
+        # which is a convoluted way of saying
+        #   theta = (1/base)^{i / d}, i \in range(0, dim, 2)
+        freqs = 1.0 / (self.rope_theta ** (torch.arange(0, dim, 2).float() / dim))
+        return freqs
+
+    def compute_sincos_cache(
+        self, position_ids: torch.Tensor, dtype: torch.dtype, device: torch.device
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        """
+        Precompute a sin/cos cache based on position_ids. This cache can
+        generally be used across. We could also rely on the compiler to do CSE,
```

Comment:
used across what?

Reply:
should be across different transformer layers